src/vm.rs

```diff
@@ -0,0 +1,74 @@
+use crate::{
+    instructions::{Add, Instruction},
+    memory::Memory,
+    opcodes::Opcode,
+    registers::{Register, Registers},
+};
+
+pub struct VM {
+    registers: Registers,
+    memory: Memory,
+}
+
+impl VM {
+    pub fn new(registers: Registers, memory: Memory) -> Self {
+        Self { registers, memory }
+    }
+
+    fn fetch(&mut self) -> u16 {
+        let mar = self.registers.get(Register::PC);
+        self.registers.program_counter_increment();
+
+        self.memory.read(mar)
+    }
+
+    fn decode(&self, bits: u16) -> Box<dyn Instruction> {
+        let opcode = Opcode::try_from(bits).unwrap();
+
+        match opcode {
+            Opcode::ADD => Box::new(Add::from_bits(bits)),
```

Comment:
If you'd still like to define `Instruction` as a `trait`, maybe you could just make `from_bits` return a `Box<dyn Instruction>` itself. That way it doesn't need to be `: Sized` and you may actually just implement the `enum` matching inside or something like that.

Reply:
Finally, I decided to use an `enum` insteaf of a `trait`.